pdc/apps/utils/middleware.py

```diff
@@ -27,3 +34,50 @@ def process_response(self, request, response):
                     messenger.send_message(topic=topic, msg=msg)
         request._messagings = None
         return response
+
+
+class ResourceCollectingMiddleware(object):
+    API_WITH_NO_PERMISSION_CONTROL = set(['changesets', 'rpc/where-to-file-bugs', 'images', 'release-types',
+                                          'variant-types', 'release-variant-types', 'content-delivery-repo-families',
+                                          'content-delivery-content-categories', 'content-delivery-content-formats',
+                                          'content-delivery-services', 'current-user-resource-permissions'])
+
+    def __init__(self):
+        action_to_obj_dict = {}
+
+        for action in ('update', 'create', 'delete', 'read'):
+            action_to_obj_dict[action] = ActionPermission.objects.get(name=action)
+
+        for prefix, view_set, basename in router.registry:
+            if prefix in self.API_WITH_NO_PERMISSION_CONTROL:
+                continue
+            resource_obj, created = Resource.objects.get_or_create(name=prefix, view=str(view_set))
+            if created:
+                logger.info("Created resource %s" % prefix)
+            for name, method in inspect.getmembers(view_set, predicate=inspect.ismethod):
+                if name.lower() in ['update', 'create', 'destroy', 'list', 'partial_update', 'retrieve']:
+                    action_permission = action_to_obj_dict[self._convert_method_to_action(name.lower())]
+                    _, created = ResourcePermission.objects.get_or_create(resource=resource_obj,
+                                                                          permission=action_permission)
+                    if created:
+                        logger.info("Created resource permission item with resource '%s' and action '%s" %
+                                    (prefix, action_permission.name))
+        raise django.core.exceptions.MiddlewareNotUsed
+
+    def _convert_method_to_action(self, method):
+        return {'update': 'update',
+                'partial_update': 'update',
+                'list': 'read',
+                'retrieve': 'read',
+                'create': 'create',
+                'destroy': 'delete'}.get(method)
+
+
+class RestrictAdminMiddleware(object):
+    """
+    Restricts access to the admin page to only logged-in users with a certain user-level.
+    """
+    def process_request(self, request):
+        if request.user.is_authenticated() and request.path.startswith(reverse('admin:index')):
+            if not (request.user.is_active and request.user.is_superuser):
+                return HttpResponseRedirect("/%s%s/" % (settings.REST_API_URL, settings.REST_API_VERSION))
```

Comment:
I thought if we grant all model permissions to all groups, then API permissions could work all by itself. PDC will not have the situation that it grant a group a API's permission, but still can't work because lack of model permission. But it will bring the risk that user could manipulate model directly to avoid permission control.

I think you mentioned we should remove model permissions. But I am not sure what it means.  Did you mean remove `django.contrib.auth.backends.ModelBackend` from `AUTHENTICATION_BACKENDS`? I tested to remove it and it looks we can't login admin page.